src/cmap/auth/mongodb_oidc/azure_machine_workflow.ts

```diff
@@ -0,0 +1,70 @@
+import { MongoAzureError } from '../../../error';
+import { request } from '../../../utils';
+import type { MongoCredentials } from '../mongo_credentials';
+import { type AccessToken, MachineWorkflow } from './machine_workflow';
+
+/** Base URL for getting Azure tokens. */
+const AZURE_BASE_URL = 'http://169.254.169.254/metadata/identity/oauth2/token?';
+
+/** Azure request headers. */
+const AZURE_HEADERS = Object.freeze({ Metadata: 'true', Accept: 'application/json' });
+
+/** Invalid endpoint result error. */
+const ENDPOINT_RESULT_ERROR =
+  'Azure endpoint did not return a value with only access_token and expires_in properties';
+
+/** Error for when the token audience is missing in the environment. */
+const TOKEN_RESOURCE_MISSING_ERROR =
+  'TOKEN_RESOURCE must be set in the auth mechanism properties when ENVIRONMENT is azure.';
+
+/**
+ * Device workflow implementation for Azure.
+ *
+ * @internal
+ */
+export class AzureMachineWorkflow extends MachineWorkflow {
+  /**
+   * Get the token from the environment.
+   */
+  async getToken(credentials?: MongoCredentials): Promise<AccessToken> {
+    const tokenAudience = credentials?.mechanismProperties.TOKEN_RESOURCE;
+    const username = credentials?.username;
+    if (!tokenAudience) {
+      throw new MongoAzureError(TOKEN_RESOURCE_MISSING_ERROR);
+    }
+    const response = await getAzureTokenData(tokenAudience, username);
```

Comment:
(this applies to all mechanisms)

From the credential caching section of the spec:

> Drivers MUST ensure that only one call to the configured provider or OIDC callback can happen at a time. 

I don't see a mechanism for preventing concurrent requests to the provider.  Would caching the promise work?  We could cache the promise in the abstract class:

```typescript
class MachineWorkflow {

    private promise = null;
	getToken() {
      this.promise ??= this._getToken();
	  try {
        return await this.promise;
	  } finally {
		this.promise = null;
	  }
	}

	abstract _getToken() {} // subclasses actually implement this method, the parent class caches the promise to ensure only one request at a time.
}
```

Reply:
This is now implemented with the `withLock` function.

Reply:
Spoke with @baileympearson I see we are intentionally making sure not to share callback results via a cached promise. He mentioned confirming that was the intention. 

IIUC maxConnecting already reduces the amount of possible parallelization (of course we're supposed to prevent it completely) is it the case that we want to ensure that the maxConnecting(default=2) connections _each_ get a unique ("unique" as in instance not necessarily as in value) result from the callback?